components/core/src/clp/ffi/ir_stream/Serializer.cpp

```diff
@@ -59,6 +258,207 @@ auto Serializer<encoded_variable_t>::change_utc_offset(UtcOffset utc_offset) ->
     serialize_utc_offset_change(m_curr_utc_offset, m_ir_buf);
 }
 
+template <typename encoded_variable_t>
+auto Serializer<encoded_variable_t>::serialize_msgpack_map(msgpack::object const& msgpack_map
+) -> bool {
+    if (msgpack::type::MAP != msgpack_map.type) {
+        return false;
+    }
+
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-union-access)
+    auto const& map{msgpack_map.via.map};
+    if (0 == map.size) {
+        serialize_empty_object(m_ir_buf);
+        return true;
+    }
+
+    m_schema_tree.take_snapshot();
+    m_schema_tree_node_buf.clear();
+    m_key_group_buf.clear();
+    m_value_group_buf.clear();
+
+    bool failure{false};
+    vector<MsgpackMapNode> working_stack;
+    working_stack.emplace_back(SchemaTree::cRootId, map.ptr, static_cast<size_t>(map.size));
+    while (false == working_stack.empty()) {
```

Comment:
Instead of iterating over each value of each map, wouldn't it make more sense to to have a nested loop as follows:

* For each map in the stack:
  * For each kv-pair in the map:
    * If the value is a primitive, encode it directly.
    * If the value is a map:
      * If it's empty, encode it directly
      * Else, add it to the stack

Reply:
Discussed offline. We decided to stick with DFS but add comments to explain what the code is doing.